packages/react-router/tests/link.test.tsx

```diff
@@ -1,44 +1,805 @@
+
+    const PostsComponent = () => {
+      const data = useLoaderData({ strict: false })
+      return (
+        <React.Fragment>
+          <h1>Posts</h1>
+          <span>Page: {data.pageDoubled}</span>
+        </React.Fragment>
+      )
+    }
+
+    const postsRoute = createRoute({
+      getParentRoute: () => rootRoute,
+      path: 'posts',
+      loaderDeps: (opts) => ({ page: opts.search }),
+      loader: loader,
+      validateSearch: (input: Record<string, unknown>) => {
+        const page = Number(input.page)
+
+        if (isNaN(page)) throw Error('Not a number!')
+
+        return {
+          page,
+        }
+      },
+      component: PostsComponent,
+    })
+
+    const router = createRouter({
+      routeTree: rootRoute.addChildren([indexRoute, postsRoute]),
+    })
+
+    render(<RouterProvider router={router} />)
+
+    const postsLink = await screen.findByRole('link', { name: 'Posts' })
+
+    expect(postsLink).toHaveAttribute('href', '/posts?page=2')
+
+    fireEvent.click(postsLink)
+
+    expect(await screen.findByText('Page: 4')).toBeInTheDocument()
+
+    expect(loader).toHaveBeenCalledOnce()
+  })
+
+  test('when navigating to /posts with a loader that errors', async () => {
+    const onError = vi.fn()
+    const rootRoute = createRootRoute()
+    const indexRoute = createRoute({
+      getParentRoute: () => rootRoute,
+      path: '/',
+      component: () => {
+        return (
+          <React.Fragment>
+            <h1>Index</h1>
+            <Link to="/posts" search={{ page: 2 }}>
+              Posts
+            </Link>
+          </React.Fragment>
+        )
+      },
+    })
+
+    const PostsComponent = () => {
+      const loader = useLoaderData({ strict: false })
+      return (
+        <React.Fragment>
+          <h1>Posts</h1>
+          <span>Page: {loader.pageDoubled}</span>
+        </React.Fragment>
+      )
+    }
+
+    const postsRoute = createRoute({
+      getParentRoute: () => rootRoute,
+      path: 'posts',
+      loaderDeps: (opts) => ({ page: opts.search }),
+      loader: () => {
+        throw new Error()
+      },
+      onError,
+      errorComponent: () => <span>Something went wrong!</span>,
+      validateSearch: (input: Record<string, unknown>) => {
+        const page = Number(input.page)
+
+        if (isNaN(page)) throw Error('Not a number!')
+
+        return {
+          page,
+        }
+      },
+      component: PostsComponent,
+    })
+
+    const router = createRouter({
+      routeTree: rootRoute.addChildren([indexRoute, postsRoute]),
+    })
+
+    render(<RouterProvider router={router} />)
+
+    const postsLink = await screen.findByRole('link', { name: 'Posts' })
+
+    expect(postsLink).toHaveAttribute('href', '/posts?page=2')
+
+    fireEvent.click(postsLink)
+
+    expect(await screen.findByText('Something went wrong!'))
+
+    expect(onError).toHaveBeenCalledOnce()
+  })
+
+  test('when navigating to /posts with a beforeLoad that redirects', async () => {
```

Comment:
In the future perhaps, this should be moved to `redirects.tests.tsx` and be tested at runtime using both `<Link>` and a button wired up to `useNavigate()`.

Reply:
I guess its hard to know where it truly lives. But I wanted to test the user pressing a link and it redirecting them because I think that's how it's actually used